slatedb/src/dispatcher.rs

```diff
@@ -457,11 +366,228 @@ pub(crate) trait MessageHandler<T: Send>: Send {
+    /// * `clock`: A [SystemClock] to use for tickers and timekeeping.
+    /// * `fp_registry`: A [FailPointRegistry] to use for fail points.
+    ///
+    /// ## Returns
+    ///
+    /// The new [MessageHandlerExecutor].
+    pub(crate) fn new_with_fp_registry(
+        error_state: WatchableOnceCell<SlateDBError>,
+        clock: Arc<dyn SystemClock>,
+        fp_registry: Arc<FailPointRegistry>,
+    ) -> Self {
+        Self {
+            tasks: Mutex::new(Some(JoinMap::new())),
+            tokens: SkipMap::new(),
+            results: Arc::new(SkipMap::new()),
+            clock,
+            error_state,
+            fp_registry,
+        }
+    }
+
+    /// Spawns a new [MessageDispatcher] for the provided [MessageHandler] on the given
+    /// [Handle]. Calls [MessageDispatcher::run] on the dispatcher in a new task. When
+    /// the dispatcher returns, [MessageDispatcher::cleanup] is called.
+    ///
+    /// ## Arguments
+    ///
+    /// * `name`: The name of the dispatcher.
+    /// * `handler`: The [MessageHandler] to use for handling messages.
+    /// * `rx`: The [mpsc::UnboundedReceiver] to use for receiving messages on the
+    ///   handler's behalf.
+    /// * `handle`: The [Handle] to use for spawning the dispatcher.
+    ///
+    /// ## Returns
+    ///
+    /// - `Ok(())` if the dispatcher was successfully spawned.
+    /// - Err([SlateDBError::BackgroundTaskStarted]) if a dispatcher with the same name is already running.
+    /// - Err([SlateDBError::BackgroundTaskExecutorStarted]) if the executor is already running.
+    pub(crate) fn spawn_on<T: Send + std::fmt::Debug + 'static>(
+        &self,
+        name: String,
+        handler: Box<dyn MessageHandler<T>>,
+        rx: mpsc::UnboundedReceiver<T>,
+        handle: &Handle,
+    ) -> Result<(), SlateDBError> {
+        // hold guard for remainder of method to prevent multiple threads
+        // from starting the same task simultaneously.
+        let mut guard = self.tasks.lock();
+        let tasks = match guard.as_mut() {
+            Some(tasks) if tasks.contains_key(&name) => {
+                return Err(SlateDBError::BackgroundTaskStarted(name))
+            }
+            Some(tasks) => tasks,
+            None => return Err(SlateDBError::BackgroundTaskExecutorStarted),
+        };
+        let token = CancellationToken::new();
+        let mut dispatcher = MessageDispatcher::new_with_fp_registry(
+            handler,
+            rx,
+            self.clock.clone(),
+            token.clone(),
+            self.fp_registry.clone(),
+        );
+        self.results.insert(name.clone(), WatchableOnceCell::new());
+        self.tokens.insert(name.clone(), token);
+        let this_error_state = self.error_state.clone();
+        let this_name = name.clone();
+        let task_future = async move {
+            let run_unwind_result = AssertUnwindSafe(dispatcher.run()).catch_unwind().await;
```

Comment:
a comment about what this "Unwind" functionality is would be nice - this is a new pattern to me